db_stress_tool/expected_state.cc

```diff
@@ -196,23 +252,325 @@ Status FileExpectedStateManager::Open() {
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      return Status::OK();
+    }
+
+    // Ignore reads.
+    Status Handle(const IteratorSeekQueryTraceRecord& /* record */,
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      return Status::OK();
+    }
+
+    // Ignore reads.
+    Status Handle(const MultiGetQueryTraceRecord& /* record */,
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      return Status::OK();
+    }
+
+    // Below are the WriteBatch::Handler overrides. We could use a separate
+    // object, but it's convenient and works to share state with the
+    // `TraceRecord::Handler`.
+
+    Status PutCF(uint32_t column_family_id, const Slice& key,
+                 const Slice& value) override {
+      uint64_t key_id;
+      if (!GetIntVal(key.ToString(), &key_id)) {
+        return Status::Corruption("unable to parse key", key.ToString());
+      }
+      uint32_t value_id = GetValueBase(value);
+
+      state_->Put(column_family_id, static_cast<int64_t>(key_id), value_id,
+                  false /* pending */);
+      ++num_write_ops_;
+      return Status::OK();
+    }
+
+    Status DeleteCF(uint32_t column_family_id, const Slice& key) override {
+      uint64_t key_id;
+      if (!GetIntVal(key.ToString(), &key_id)) {
+        return Status::Corruption("unable to parse key", key.ToString());
+      }
+
+      state_->Delete(column_family_id, static_cast<int64_t>(key_id),
+                     false /* pending */);
+      ++num_write_ops_;
+      return Status::OK();
+    }
+
+    Status SingleDeleteCF(uint32_t column_family_id,
+                          const Slice& key) override {
+      return DeleteCF(column_family_id, key);
+    }
+
+    Status DeleteRangeCF(uint32_t column_family_id, const Slice& begin_key,
+                         const Slice& end_key) override {
+      uint64_t begin_key_id, end_key_id;
+      if (!GetIntVal(begin_key.ToString(), &begin_key_id)) {
+        return Status::Corruption("unable to parse begin key",
+                                  begin_key.ToString());
+      }
+      if (!GetIntVal(end_key.ToString(), &end_key_id)) {
+        return Status::Corruption("unable to parse end key",
+                                  end_key.ToString());
+      }
+
+      state_->DeleteRange(column_family_id, static_cast<int64_t>(begin_key_id),
+                          static_cast<int64_t>(end_key_id),
+                          false /* pending */);
+      ++num_write_ops_;
+      return Status::OK();
+    }
+
+    Status MergeCF(uint32_t column_family_id, const Slice& key,
+                   const Slice& value) override {
+      return PutCF(column_family_id, key, value);
+    }
+
+   private:
+    uint64_t num_write_ops_ = 0;
+    uint64_t max_write_ops_;
+    ExpectedState* state_;
+  };
+
+  SequenceNumber seqno = db->GetLatestSequenceNumber();
+  if (seqno < saved_seqno_) {
```

Comment:
For clarity, I would `assert(HasHistory())`

Reply:
Done.